In [1]:
# WEB DEV TEAM PLEASE DELETE BEFORE PUBLISHING - DS TEAM PLEASE KEEP IN THE FINAL VERSION OF THE USE CASE

from IPython.core.display import HTML
HTML("""
<style>
@import url('https://fonts.googleapis.com/css2?family=Poppins:wght@600;700&family=Inter:wght@400;500&display=swap');
body,p,li,div{font-family:'Inter',sans-serif}
body{counter-reset:h2counter}
.usecase-header-bar,.usecase-title,.usecase-section-heading,.usecase-sub-section-heading,.usecase-sub-sub-section-heading,.usecase-unnumbered-heading{font-family:'Poppins',sans-serif}
.usecase-header-bar{padding:16px 20px;background-color:#059669;color:#fff;border-radius:6px 6px 0 0}
.usecase-title{font-size:1.9em;font-weight:800}
.usecase-details{padding-left:20px;padding-bottom:8px;padding-top:8px;font-size:1.1em;color:#121212}
.usecase-section-heading,.usecase-unnumbered-heading{font-weight:700;font-size:1.35em;color:#059669;padding:14px 4px 6px 4px;border-bottom:3px solid #2DBE6C;margin-top:10px}
.usecase-section-heading::before{content:none!important}
.usecase-sub-section-heading{font-weight:600;font-size:1.15em;color:#059669;margin-top:8px}
.usecase-sub-section-heading::before{content:none!important}
.usecase-sub-sub-section-heading{font-weight:500;font-size:1.05em;color:#059669;margin-top:6px}
.usecase-sub-sub-section-heading::before{content:none!important}
.usecase-body{font-size:1em;color:#121212;line-height:1.6;margin-bottom:8px}
.usecase-list{line-height:1.7}
.usecase-list ul,.usecase-list ol{padding-left:24px;margin-top:8px;margin-bottom:12px}
.usecase-list li{margin-bottom:6px}
.usecase-datasets{background-color:rgba(5,150,105,.06);border-left:4px solid #059669;padding:12px 20px;border-radius:0 4px 4px 0;margin-top:12px}
.usecase-datasets ul{margin-top:8px;margin-bottom:8px}
.usecase-contents a,.usecase-datasets a{color:#059669;text-decoration:none}
.usecase-datasets a{font-weight:500}
.usecase-contents a:hover,.usecase-datasets a:hover{text-decoration:underline}
.usecase-contents{line-height:1.9;padding-left:20px}
</style>
""")

<div class="usecase-header-bar"><div class="usecase-title">Smart Bin Optimised Waste Collection Route</div></div>

<div class="usecase-details">
<b>Author:</b> Sachitha Sadeepa Kasthuriarachchi<br>
<b>Created:</b> Trimester 1, 2024<br>
<b>Updated:</b> Trimester 2, 2026 (Soham Chodanekar)<br>
<b>Time to Complete:</b> 180 mins<br>
<b>Level:</b> Beginner/Intermediate<br>
<b>Pre-requisite Skills:</b> Python, Jupyter Notebooks and Data Analysis
</div>

<div id="scenario" class="usecase-unnumbered-heading"><b>Scenario</b></div>

<div class="usecase-body">

As a waste management authority in Melbourne, I want to plan smart-bin collection routes that prioritise bins that require collection while reducing unnecessary travel between locations.

This will help to support more efficient planning of smart-bin collection activities across the city.

</div>

<div id="what-this-teaches" class="usecase-unnumbered-heading"><b>What this use case will teach you</b></div>

<div class="usecase-body">

At the end of this use case you will learn:
</div>

<div class="usecase-list">

- Route Optimisation: Ability to develop algorithms and methodologies for optimising waste collection routes, considering factors such as distance, time, and resource utilisation.

- Data Integration: Proficiency in integrating real-time data from smart bins into route planning algorithms, ensuring accurate and timely decision-making.

- Environmental Impact Analysis: Capability to assess and mitigate environmental impacts of waste collection operations, including reducing fuel consumption and carbon emissions.

- Cost Optimisation: Skills in cost analysis and resource allocation to minimise operational expenses associated with waste collection, including labour, fuel, and vehicle maintenance.

- Scalability Planning: Ability to design scalable solutions to accommodate the growing number of smart bins and evolving waste management needs in urban environments.

- User Experience Design: Proficiency in designing user-friendly interfaces for waste management authorities, ensuring ease of use and efficient management of collection routes.

</div>

<div id="background" class="usecase-unnumbered-heading"><b>Background</b></div>

<div class="usecase-body">

In Melbourne, managing urban waste efficiently is important for maintaining cleanliness, reducing environmental impact and supporting effective city operations. Smart-bin technology provides useful information such as bin fill levels and geographic locations, which can be used to improve how collection routes are planned. As the waste management authority in Melbourne, the aim of this use case is to improve smart-bin collection by prioritising bins that require collection and reducing unnecessary travel between locations. The analysis uses smart-bin fill levels and geographic coordinates to plan and visualise a more efficient collection route. The use case focuses on improving route planning and reducing estimated travel distance. By considering the location of the bins together with their fill levels, the analysis provides a practical comparison between a simple fill-priority route and a route that also considers proximity between stops.
</div>

<div class="usecase-list">

- Efficient Route Planning: The analysis proposes a collection route using smart-bin fill levels and geographic proximity to reduce unnecessary travel between bins.

- Use of Current Bin Data: The route planning uses the available smart-bin fill-level and location data to prioritise bins that require collection.

- Reduced Travel Distance: The proposed route is compared with a simple fill-priority baseline to evaluate whether considering proximity can reduce the estimated collection distance.


</div>

<div id="data-sets" class="usecase-unnumbered-heading"><b>Datasets</b></div>

<div class="usecase-body">

This analysis uses one smart bin sensor dataset containing bin fill levels, locations and sensor information. One dataset is sufficient for this use case because it provides the main information required to analyse bin status and plan collection routes.

</div>

<div class="usecase-datasets">

- <a href="https://discover.data.vic.gov.au/dataset/smart-bins-argyle-square/resource/07837d22-e224-40a1-a212-b8dcbcb45e98">Smart Bins Argyle Square – DataVic</a>

</div>

<div id="contents" class="usecase-unnumbered-heading"><b>Contents</b></div>

1. <a href="#Importing-Data">Importing Data</a><br>
&nbsp;<a href="#Dataset-Imported-through-API">1.1 Dataset Imported through API</a><br>
&nbsp;<a href="#Fetching-Smart-Bin-Sensor-Data">1.2 Fetching Smart Bin Sensor Data</a><br>

2. <a href="#Pre-processing-Data">Pre-processing Data</a><br>
&nbsp;<a href="#Identifying-Missing-Values">2.1 Identifying Missing Values</a><br>
&nbsp;<a href="#Optimising-Data-Types">2.2 Optimising Data Types</a><br>
&nbsp;<a href="#Improving-Time-Related-Data">2.3 Improving Time-Related Data</a><br>
&nbsp;<a href="#Identifying-Unique-Sensor-Serial-Numbers">2.4 Identifying Unique Sensor Serial Numbers</a><br>
&nbsp;<a href="#Naming-Sensors-Based-on-Serial-Numbers">2.5 Naming Sensors Based on Serial Numbers</a><br>
&nbsp;<a href="#Identifying-Missing-Values-in-the-Dataset-(Part2)">2.6 Identifying Missing Values in the Dataset (Part 2)</a><br>
&nbsp;<a href="#Cleaning-the-Dataset-by-Handling-Missing-Data">2.7 Cleaning the Dataset by Handling Missing Data</a><br>
&nbsp;<a href="#Updating-the-new-date-Column-Data-Type">2.8 Updating the new 'date' Column Data Type</a><br>
&nbsp;<a href="#Re-Checking-Current-Data-Types-in-the-Dataset">2.9 Re-checking Current Data Types in the Dataset</a><br>
&nbsp;<a href="#Analysing-Latitude-and-Longitude-Column-for-Completeness">2.10 Analysing Latitude and Longitude Column for Completeness</a><br>
&nbsp;<a href="#Inspecting-a-Non-Null-Latitude-and-Longitude-Entry">2.11 Inspecting a Non-Null Latitude and Longitude Entry</a><br>
&nbsp;<a href="#Confirming-Data-Preprocessing-with-a-New-DataFrame">2.12 Confirming Data Preprocessing with a New DataFrame</a>

3. <a href="#visualisation-part-1">Visualisation Part 1</a><br>
&nbsp;<a href="#Visualising-Geolocation-Data">3.1 Visualising Geolocation Data</a><br>
&nbsp;<a href="#Adding-Markers-to-the-Map">3.2 Adding Markers to the Map</a><br>
&nbsp;<a href="#Visualising-Sensor-Locations-on-a-Map">3.3 Visualising Sensor Locations on a Map</a>

4. <a href="#visualisation-part-2">Visualisation Part 2</a><br>
&nbsp;<a href="#Preparing-Fill-Levels">4.1 Preparing Fill Levels</a><br>
&nbsp;<a href="#Calculating-Distance-Matrix">4.2 Calculating Distance Matrix</a><br>
&nbsp;<a href="#Route-Optimisation">4.3 Route Optimisation</a><br>
&nbsp;&nbsp;&nbsp;<a href="#Prioritising-by-Fill-Level">4.3.1 Prioritising-by-Fill-Level</a><br>
&nbsp;&nbsp;&nbsp;<a href="#Route-Initialisation">4.3.2 Route-Initialisation</a><br>
&nbsp;&nbsp;&nbsp;<a href="#Determine-Next-Stop">4.3.3 Determine-Next-Stop</a><br>
&nbsp;<a href="#Visualising-the-Optimised-Route">4.4 Visualising the Optimised Route</a><br>
&nbsp;&nbsp;&nbsp;<a href="#Setting-Up-the-Map">4.4.1 Setting-Up-the-Map</a><br>
&nbsp;&nbsp;&nbsp;<a href="#Marking-Stops">4.4.2 Marking-Stops</a><br>
&nbsp;&nbsp;&nbsp;<a href="#Illustrating-the-Route">4.4.3 Illustrating-the-Route</a><br>


</div>

<div id="Importing-Data" class="usecase-section-heading"><b>1. Importing Data</b></div>

<div class="usecase-body">This section imports the required libraries and retrieves the smart bin sensor data.</div>

<div id="Dataset-Imported-through-API" class="usecase-sub-section-heading"><b>1.1 Dataset Imported through API</b></div>

In [ ]:
import requests
import pandas as pd
import numpy as np
import folium
from sklearn.metrics.pairwise import haversine_distances

In [ ]:
def fetch_data(base_url, dataset, num_records=99, offset=0):
    all_records = []
    max_offset = 9900  # Maximum number of requests

    while True:
        # maximum limit check
        if offset > max_offset:
            break

        # Create API request URL
        filters = f'{dataset}/records?limit={num_records}&offset={offset}'
        url = f'{base_url}{filters}'

        # Start request
        try:
            result = requests.get(url, timeout=10)
            result.raise_for_status()
            records = result.json().get('results')
        except requests.exceptions.RequestException as e:
            raise Exception(f"API request failed: {e}")
        if records is None:
            break
        all_records.extend(records)
        if len(records) < num_records:
            break

        # next cycle offset
        offset += num_records

    # DataFrame all data
    df = pd.DataFrame(all_records)
    return df    


<div id="Fetching-Smart-Bin-Sensor-Data" class="usecase-sub-section-heading"><b>1.2 Fetching Smart Bin Sensor Data</b></div>

<div class="usecase-body">

This step retrieves the smart bin sensor dataset from the Victorian Government DataVic API.

</div>

<div class="usecase-list">

- **Define the API endpoint:** The DataVic `datastore_search` endpoint is used to access the dataset.

- **Specify the resource:** The dataset is identified using its unique `resource_id`.

- **Fetch the data:** The records are requested from the API and converted into a pandas DataFrame.
</div>
<div class="usecase-body">
This allows the smart-bin sensor data to be loaded directly into the notebook for further cleaning and analysis.
</div>


In [3]:
url = "https://discover.data.vic.gov.au/en_AU/api/3/action/datastore_search"

params = {
    "resource_id": "07837d22-e224-40a1-a212-b8dcbcb45e98",
    "limit": 30631
}

response = requests.get(url, params=params, headers={"User-Agent": "Mozilla/5.0"}, timeout=30)
response.raise_for_status()

records = response.json()["result"]["records"]

data = pd.DataFrame(records)

data.head(3)



,_id,time,serial,sensor_name,status_current_fill_level,bin_status,last_update,product_type,description,address,latlong,wastebasket_size
0,1,2021-05-15T09:01:49+00:00,FB1000001905AA12,Internal Bin Sensor 5,19,Low,None,CleanFLEX,Argyle - FB1000001905AA12,"Piazza Italia, Unnamed Road, Carlton VIC 3053,...","-37.802367, 144.966167",None
1,2,2021-05-15T10:53:51+00:00,FB1000001905AA11,None,23,Low,None,CleanFLEX,Argyle - FB1000001905AA11,"10 Argyle Pl N, Carlton VIC 3053, Australia","-37.802133, 144.966083",None
2,3,2021-05-16T02:06:58+00:00,FB1000001905AA07,Internal Bin Sensor 7,5,Low,None,CleanFLEX,Argyle - FB1000001905AA07,"Grattan Pl, Carlton VIC 3053, Australia","-37.803117, 144.965233",None


<div id="Pre-processing-Data" class="usecase-section-heading"><b>2. Pre-processing Data</b></div>

<div class="usecase-body"> This section prepares and cleans the sensor data for analysis and visualisation. </div>

<div id="Identifying-Missing-Values" class="usecase-sub-section-heading"><b>2.1 Identifying Missing Values in the Dataset (Part 1)</b></div>
<div class="usecase-body">
To ensure the integrity and quality of our analysis, it's crucial to check for missing values within our dataset. The code performs the following steps:

To check the quality of the dataset, we first identify any missing values. This helps us understand which columns may need cleaning before continuing with the analysis.
</div>

<div class="usecase-list">

- **Calculate Missing Values:** `data.isnull().sum()` is used to count missing values in each column.
- **Display Missing Values:** The results are printed so we can quickly see which columns contain missing data.
</div>

<div class="usecase-body">
This check helps us decide what cleaning steps are needed before using the data for further analysis.
</div>
<div class="usecase-body">Understanding where our data is incomplete is crucial for deciding on subsequent data cleaning and imputation strategies, ensuring the reliability of our analysis or model predictions.</div>


In [4]:
missing_values = data.isnull().sum()
print(" Checking for the Missing Values ")
print(missing_values)

 Checking for the Missing Values 
_id                              0
time                             0
serial                           0
sensor_name                   8921
status_current_fill_level        0
bin_status                       0
last_update                  30631
product_type                     0
description                      0
address                          0
latlong                       5585
wastebasket_size             30631
dtype: int64


<div id="Optimising-Data-Types" class="usecase-sub-section-heading"><b>2.2 Optimising Data Types for Analysis</b></div>

<div class="usecase-body">
To ensure our dataset is primed for analysis, we undertake a comprehensive data type optimisation process. This involves adjusting the data types of various columns to more accurately reflect their content and optimise memory usage. Here's an overview of the steps taken:
</div>
<div class="usecase-list">

- **Timestamp Conversion**: Convert the `time` and `last_update` columns to `datetime` objects to facilitate time series analysis.
- **String Specification**: Change `serial`, `sensor_name`, `description`, and `address` columns to the `string` data type for consistency in text data handling.
- **Category Optimisation**: For columns like `bin_status` and `product_type`, we use the `category` data type when we have a limited set of possible values, enhancing memory efficiency and performance during grouping operations.
- **Numerical Adjustments**: Ensure numerical data, such as `wastebasket_size`, is correctly typed (`float`) for accurate mathematical computations.
- **Geolocation Handling**: Split the `Latlong` column into separate `latitude` and `longitude` columns and convert them to `float` data types, dropping the original `Latitude and Longitude` column to simplify geographical analysis.

These adjustments are crucial for the following reasons:
- **Efficiency**: Optimising data types can significantly reduce memory usage, making data manipulation more efficient.
- **Accuracy**: Ensuring data types match the nature of the data prevents errors during analysis.
- **Convenience**: Having the right data type (e.g., datetime) simplifies operations like date range filtering or time-based aggregation.
</div>

<div class="usecase-body"> By refining our dataset's structure through these changes, we lay a solid foundation for robust and error-free data analysis. </div>


In [5]:
# Let's change the datatypes to more appropriate types:
data['time'] = pd.to_datetime(data['time'])
data['serial'] = data['serial'].astype('string')
data['sensor_name'] = data['sensor_name'].astype('string')
# 'status_current_fill_level' is already an integer, no change needed.
data['bin_status'] = data['bin_status'].astype('category')  # Or 'string' if it's not a finite set of statuses.
data['last_update'] = pd.to_datetime(data['last_update'])
data['product_type'] = data['product_type'].astype('category')  # Or 'string' if it's not a finite set of types.
data['description'] = data['description'].astype('string')
data['address'] = data['address'].astype('string')
# Assuming 'wastebasket_size' refers to a numerical measurement and not a category.
data['wastebasket_size'] = data['wastebasket_size'].astype('float')  # Or 'category' if it's not numerical.
# Split the latlong column into separate latitude and longitude columns
# Convert both values to numeric format for geographic analysis.
if not data.empty and 'latlong' in data.columns and data['latlong'].str.contains(',').any():
    latlong_split = data['latlong'].str.split(',', expand=True)

    data['latitude'] = pd.to_numeric(latlong_split[0], errors='coerce')
    data['longitude'] = pd.to_numeric(latlong_split[1], errors='coerce')

    data.drop('latlong', axis=1, inplace=True)
# Display the DataFrame with new datatypes
data.dtypes


_id                                        int64
time                         datetime64[ns, UTC]
serial                            string[python]
sensor_name                       string[python]
status_current_fill_level                 object
bin_status                              category
last_update                       datetime64[ns]
product_type                            category
description                       string[python]
address                           string[python]
wastebasket_size                         float64
latitude                                 float64
longitude                                float64
dtype: object

<div id="Improving-Time-Related-Data" class="usecase-sub-section-heading"><b>2.3 Improving the Detail of Time-Related Data</b></div>

<div class="usecase-body">
This segment focuses on refining our dataset's temporal data by splitting the 'time' column into distinct 'date' and 'time_only' columns. This transformation aids in detailed time-based analysis and enables more flexible data manipulation. Here are the key steps involved:
</div>

<div class="usecase-list">

- **Datetime Parsing**: Initially, we ensure the 'time' column is correctly parsed into datetime format with `pd.to_datetime(data['time'], errors='coerce')`. The `errors='coerce'` parameter converts any parsing errors into NaT (Not a Time), preventing data corruption.

- **Column Creation**: Post successful datetime conversion, we proceed to split the datetime column:
   - The `date` column is created to store only the date component of each entry.
   - The `time_only` column is created to store only the time component.

- **Error Handling**: In cases where datetime parsing results in all null values (indicating parsing issues), both `date` and `time_only` columns are set to `None`. This precaution ensures the dataset's integrity remains uncompromised.

- **Previewing Changes**: Finally, we display the first few rows of the newly created `date` and `time_only` columns to verify the split was successful and to provide a glimpse into the data structure post-transformation.
</div>
<div class="usecase-body"> This enhancement is pivotal for conducting in-depth temporal analysis and offers the flexibility to analyse trends and patterns based on specific dates or times. </div>



In [6]:
# Splitting the 'time' column into separate 'date' and 'time' columns, this time ensuring to handle any potential issues
data['time'] = pd.to_datetime(data['time'], errors='coerce')  # Parsing the datetime to ensure correct format

# If datetime parsing was successful, create separate 'date' and 'time' columns
if not data['time'].isnull().all():
    data['date'] = data['time'].dt.date
    data['time_only'] = data['time'].dt.time
else:
    data['date'], data['time_only'] = [None, None]

placeholder_output_structure_2 = data[['date', 'time_only']].head(3)
placeholder_output_structure_2 

,date,time_only
0,2021-05-15,09:01:49
1,2021-05-15,10:53:51
2,2021-05-16,02:06:58


<div id="Identifying-Unique-Sensor-Serial-Numbers" class="usecase-sub-section-heading"><b>2.4 Identifying Unique Sensor Serial Numbers</b></div>
 
<div class="usecase-body">
An essential step in our data analysis process involves identifying the unique serial numbers of the sensors within our dataset. This enables us to differentiate and potentially name each sensor for more detailed and personalised analysis. Here's the procedure:
</div>

<div class="usecase-list">

- **Retrieve Unique Serials**: We use `data['serial'].unique()` to extract an array of unique serial numbers from the `serial` column of our dataset. The `.unique()` method is instrumental in identifying all distinct sensors present, ensuring no duplicates are considered.

- **Displaying Unique Serials**: The resulting array of unique serial numbers is then displayed. This array is crucial for subsequent steps in our analysis, such as naming sensors, analysing data on a per-sensor basis, or grouping data for sensor-specific insights.
</div>

<div class="usecase-body">This step is fundamental in understanding the diversity and range of sensors within our dataset, laying the groundwork for targeted sensor-specific analyses and operations.
</div>


In [7]:
missing_values_count = data['sensor_name'].isnull().sum()
print(f"Number of missing values in column_name: {missing_values_count}")


Number of missing values in column_name: 8921


In [8]:
unique_serial_numbers = data['serial'].unique()
unique_serial_numbers

<StringArray>
['FB1000001905AA12', 'FB1000001905AA11', 'FB1000001905AA07',
 'FB1000001905AA16', 'FB1000001905AA08', 'FB1000001905AA13',
 'FB1000001905AA17', 'FB1000001905AA02', 'FB1000001905AA18',
 'FB1000001905AA00', 'FB1000001905AA04', 'FB1000001905AA05',
 'FB1000001905AA15', 'FB1000001905AA19', 'FB1000001905AA14',
 'FB1000001905AA01', 'FB1000001905AA06', 'FB1000001905AA03',
 'FB1000001905AA09', 'FB1000001905AA10']
Length: 20, dtype: string

<div id="Naming-Sensors-Based-on-Serial-Numbers" class="usecase-sub-section-heading"><b>2.5 Naming Sensors Based on Serial Numbers</b></div>
<div class="usecase-body">
To enhance clarity and facilitate easier identification of sensors in our dataset, we adopt a naming convention that incorporates the last two digits of each sensor's serial number. This step not only simplifies the analysis but also adds a layer of readability when dealing with multiple sensors. Here's how we accomplish this:
</div>

<div class="usecase-list">

- **Update Sensor Names**: We prefix every sensor name with "Sensor-" and append the last two digits of its serial number using `data['serial'].str[-2:]`. This approach ensures each sensor has a unique and identifiable name based on its serial number.

- **Reflecting Changes**: After updating the 'sensor_name' column, we create a placeholder DataFrame to preview the changes. This DataFrame includes both 'serial' and 'sensor_name' columns, displaying how the sensor names have been updated based on the serial numbers.

- **Preview Output**: Using `.head()` on our placeholder DataFrame, we show the first few rows to verify the naming convention's application and ensure the data integrity is maintained.
</div>

<div class="usecase-body">This naming strategy not only aids in the quick identification of sensors but also adds a systematic approach to managing sensor data, making subsequent data analysis tasks more intuitive. </div>


In [9]:
data['sensor_name'] = "Sensor-" + data['serial'].str[-2:]
placeholder_output_structure_1 = data[['serial', 'sensor_name']]  
placeholder_output_structure_1.head(3) 

,serial,sensor_name
0,FB1000001905AA12,Sensor-12
1,FB1000001905AA11,Sensor-11
2,FB1000001905AA07,Sensor-07


In [10]:
missing_values_count = data['sensor_name'].isnull().sum()
print(f"Number of missing values in column_name: {missing_values_count}")

Number of missing values in column_name: 0


<div id="Identifying-Missing-Values-in-the-Dataset-(Part2)" class="usecase-sub-section-heading"><b> 2.6 Identifying Missing Values in the Dataset (Part 2)</b></div>

<div class="usecase-body">
A crucial step in data preprocessing involves identifying and quantifying missing values within our dataset. This process enables us to understand the extent of missing data and strategize on how to handle it. We approach this in two main steps:
</div>
<div class="usecase-list">

- **Identifying Missing Values**: We utilise `data.isnull().sum()` to calculate the total number of missing values in each column of our dataset. This method checks each cell for null values, providing a column-wise sum. The output is printed under "Current Missing Values," offering immediate visibility into which columns contain missing data and to what extent.

- **Calculating Missing Data Proportion**: To grasp the significance of the missing values, we calculate their proportion relative to the total dataset size. This is achieved by dividing the number of missing values in each column by the total number of rows in the dataset, then multiplying by 100 to convert it to a percentage. The results are printed under "Current Missing Values Proportion," giving us a clear picture of how much data is missing across each column.
</div>

<div class="usecase-body">This assessment not only highlights the areas of the dataset that may require attention due to missing data but also informs our decision-making process regarding data cleaning techniques, such as imputation or removal, ensuring the robustness of our analysis.</div>



In [11]:
missing_values = data.isnull().sum()
print("Current Missing Values")
print(missing_values)

# Calculation of the missing proportion
missing_proportion = (missing_values / len(data)) * 100

print("Current Missing Values Proportion:")
print(missing_proportion)


Current Missing Values
_id                              0
time                             0
serial                           0
sensor_name                      0
status_current_fill_level        0
bin_status                       0
last_update                  30631
product_type                     0
description                      0
address                          0
wastebasket_size             30631
latitude                      5585
longitude                     5585
date                             0
time_only                        0
dtype: int64
Current Missing Values Proportion:
_id                            0.000000
time                           0.000000
serial                         0.000000
sensor_name                    0.000000
status_current_fill_level      0.000000
bin_status                     0.000000
last_update                  100.000000
product_type                   0.000000
description                    0.000000
address                        0.000000
wast

<div id="Cleaning-the-Dataset-by-Handling-Missing-Data" class="usecase-sub-section-heading"><b>2.7 Cleaning the Dataset by Handling Missing Data</b></div>

<div class="usecase-body">
A vital part of preparing our dataset for analysis involves ensuring that it does not contain significant gaps that could distort our findings. To this end, we take two decisive actions to handle missing data effectively:
</div>

<div class="usecase-list">

- **Dropping Rows with Missing `Latitude and Longitude` Data**: Recognising the critical nature of geographical data for our analysis, we remove any rows where the `Latitude and Longitude` information is missing. This is achieved with `data.dropna(subset=[Latitude and Longitude])`, ensuring that our dataset only includes entries with valid geolocation data. This step is crucial for analyses that depend on spatial information.

- **Removing Unnecessary Columns**: Further refining our dataset, we decide to eliminate columns that are not essential to our specific analytical goals. In this case, `last_update` and `wastebasket_size` columns are removed using `data.drop(columns=['last_update', 'wastebasket_size'])`. This decision might be driven by factors such as the columns' relevance to the research question, a high proportion of missing values, or the desire to streamline the dataset for efficiency.
</div>

<div class="usecase-body">By taking these steps, we ensure our dataset is more focused and manageable, eliminating potential sources of error or bias that could arise from missing or irrelevant data. This process sets a strong foundation for accurate and meaningful analysis.</div>


In [12]:
print(data.columns.tolist())

['_id', 'time', 'serial', 'sensor_name', 'status_current_fill_level', 'bin_status', 'last_update', 'product_type', 'description', 'address', 'wastebasket_size', 'latitude', 'longitude', 'date', 'time_only']


In [13]:
#Dropping rows with missing latitude and longitude
data = data.dropna(subset=['latitude', 'longitude'])


<div id="Updating-the-new-date-Column-Data-Type" class="usecase-sub-section-heading"><b>2.8 Updating the new 'date' Column Data Type</b></div>
<div class="usecase-body">
To ensure our dataset's `date` column is in the optimal format for date-related operations and analysis, we convert it to a datetime data type. This conversion is crucial for several reasons:
</div>

<div class="usecase-list">

- **Consistency and Accuracy**: Converting the `date` column to a datetime format using `pd.to_datetime(data['date'])` ensures that all entries are consistently recognised as dates. This uniformity is vital for accurate sorting, filtering, and time series analysis.

- **Enhanced Date Functionality**: With the `date` column in datetime format, we unlock a range of Pandas functionalities specific to date and time operations. This includes extracting components of the date (like the month or year), performing date arithmetic, and efficiently handling time zones if necessary.
</div>

<div class="usecase-body">This step is a fundamental part of data preprocessing, setting the stage for any analysis or operations that rely on temporal data. It ensures our dataset is accurately prepared to support insightful and reliable temporal analyses.</div>


In [14]:
data['date'] = pd.to_datetime(data['date'])

<div id="Re-Checking-Current-Data-Types-in-the-Dataset" class="usecase-sub-section-heading"><b>2.9 Re-checking Current Data Types in the Dataset</b></div>

<div class="usecase-body">
After performing various data preprocessing steps, including handling missing values and optimising data types, it's important to review the current state of our dataset's structure. This involves checking the data types of each column to ensure they align with our expectations and requirements for analysis. Here's how we do it: </div>

<div class="usecase-list">

- **Print Statement**: We use a simple print statement to introduce the context, stating "Current data types".

- **Displaying Data Types**: The `data.dtypes` command is executed to list each column's data type within our DataFrame. This output is crucial for verifying that our data preprocessing steps have been successfully applied, ensuring that each column is now properly formatted for efficient and accurate analysis.
</div>

<div class="usecase-body"> By reviewing the data types, we confirm the dataset's readiness for further analysis, ensuring that numerical data can be correctly manipulated and that categorical and date-time information is appropriately formatted. </div>



In [15]:
print("Current data types")
print(data.dtypes)

Current data types
_id                                        int64
time                         datetime64[ns, UTC]
serial                            string[python]
sensor_name                       string[python]
status_current_fill_level                 object
bin_status                              category
last_update                       datetime64[ns]
product_type                            category
description                       string[python]
address                           string[python]
wastebasket_size                         float64
latitude                                 float64
longitude                                float64
date                              datetime64[ns]
time_only                                 object
dtype: object


<div id="Analysing-Latitude-and-Longitude-Column-for-Completeness" class="usecase-sub-section-heading"><b>2.10 Analysing Latitude and Longitude Column for Completeness</b></div>

<div class="usecase-body">
Ensuring the completeness and accuracy of geolocation data in our Latitude and Longitude column is crucial for spatial analysis. To this end, we conduct a two-fold examination to identify any missing or blank entries: </div>

<div class="usecase-list">

- **Missing Values Check**: We calculate the total number of missing (NaN) values in the Latitude and Longitude column using `data[Latitude and Longitude].isna().sum()`. This step helps us understand if there are any entries that lack geolocation data entirely, which could impact geospatial analyses and visualisations.

- **Blank Cells Check**: To identify entries that might appear non-empty but contain only blank spaces, we use `data[Latitude and Longitude].str.strip().eq('').sum()`. This process strips any leading and trailing spaces from each entry before comparing it to an empty string, allowing us to count how many cells are effectively blank despite not being NaN.
</div>

<div class="usecase-body"> The results of these checks are printed, providing insight into the data quality of the Latitude and Longitude column. Identifying and quantifying both missing and blank values are essential steps in data cleaning, ensuring that subsequent spatial analyses are based on complete and accurate data. </div>



In [16]:
# Check for missing values in Latitude and Longitude column
missing_values_count = data[['latitude', 'longitude']].isna().sum()
print(f"Missing values in 'latitude', 'longitude': {missing_values_count}")
 
# Check for blank cells in Latitude and Longitude column
blank_cells_count = data[['latitude', 'longitude']].eq('').sum()
print(f"Blank cells in 'latitude', 'longitude': {blank_cells_count}")

Missing values in 'latitude', 'longitude': latitude     0
longitude    0
dtype: int64
Blank cells in 'latitude', 'longitude': latitude     0
longitude    0
dtype: int64


<div id="Inspecting-a-Non-Null-Latitude-and-Longitude-Entry" class="usecase-sub-section-heading"><b>2.11 Inspecting a Non-Null Latitude and Longitude Entry</b></div>
<div class="usecase-body">
To better understand the format and structure of our geolocation data within the Latitude and Longitude column, we inspect a sample non-null entry. This examination is pivotal for planning any required data transformation or extraction steps, especially when preparing for spatial analysis. Here's our approach: </div>

<div class="usecase-list">

- **Filtering Non-Null Entries**: We start by removing all null entries from the Latitude and Longitude column using `data[Latitude and Longitude].dropna()`, ensuring we're only working with valid geolocation data.

- **Sample Inspection**: If the filtered series is not empty, indicating the presence of non-null geolocation data, we print the first non-null entry using `.iloc[0]`. This provides us with a clear example of the data format we're dealing with.

- **Handling Absence of Data**: In cases where no non-null entries exist within the Latitude and Longitude column, we print a message stating "No non-null entries found in Latitude and Longitude." This helps in identifying datasets that may lack the necessary geolocation data for spatial analysis.
</div>

<div class="usecase-body">Inspecting a sample entry allows us to understand the formatting of our geolocation data, informing any necessary preprocessing steps to extract latitude and longitude for analysis purposes. </div>


In [17]:
# Print the structure of a non-null 'lat_long' entry
non_null_entries = data[['latitude', 'longitude']].dropna()
if not non_null_entries.empty:
    print(non_null_entries.iloc[0])
else:
    print("No non-null entries found in 'latitude', 'longitude'.")

latitude     -37.802367
longitude    144.966167
Name: 0, dtype: float64


<div id="Confirming-Data-Preprocessing-with-a-New-DataFrame" class="usecase-sub-section-heading"><b>2.12 Confirming Data Preprocessing with a New DataFrame </b></div>
<div class="usecase-body">
After meticulous data cleaning and preprocessing, including handling missing values, optimising data types, and extracting critical geolocation information, we consolidate our efforts into a new DataFrame. This new DataFrame serves as a snapshot to confirm the effective application of our preprocessing steps. The process involves the following:
</div>

<div class="usecase-list">

- **Selection of Key Columns**: We create `new_DF` by selecting columns that are essential for our subsequent analyses. These columns include 'date', 'time_only', 'serial', 'sensor_name', 'status_current_fill_level', 'latitude', and 'longitude'. This selection not only reflects the data cleaning and transformation efforts undertaken but also aligns with our analysis objectives by focusing on crucial data points.

- **DataFrame Creation**: The new DataFrame, `new_DF`, is instantiated with the chosen columns, providing a streamlined and focused dataset for analysis. This step is crucial for verifying that our preprocessing tasks have been successfully applied and that the dataset is now in a suitable format for analysis.

- **Review and Verification**: Displaying `new_DF` allows us to visually confirm the presence and correct formatting of the selected columns, ensuring that our dataset is ready for the next stages of our project.
</div>

<div class="usecase-body"> This consolidation into a new DataFrame is a critical step in our data preparation process, enabling us to proceed with confidence in the quality and relevance of our data for analysis. </div>


In [18]:
new_DF = data[['date', 'time_only','serial','sensor_name','status_current_fill_level','latitude','longitude']]
new_DF.head(3) 

,date,time_only,serial,sensor_name,status_current_fill_level,latitude,longitude
0,2021-05-15,09:01:49,FB1000001905AA12,Sensor-12,19,-37.802367,144.966167
1,2021-05-15,10:53:51,FB1000001905AA11,Sensor-11,23,-37.802133,144.966083
2,2021-05-16,02:06:58,FB1000001905AA07,Sensor-07,5,-37.803117,144.965233


In [19]:
new_DF['latitude'] = pd.to_numeric(
    new_DF['latitude'],
    errors='coerce'
)

new_DF['longitude'] = pd.to_numeric(
    new_DF['longitude'],
    errors='coerce'
)

new_DF['status_current_fill_level'] = pd.to_numeric(
    new_DF['status_current_fill_level'],
    errors='coerce'
)

new_DF = new_DF.dropna(
    subset=['latitude', 'longitude']
)

new_DF.head(3)

C:\Users\Soham\AppData\Local\Temp\ipykernel_34712\1807732926.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_DF['latitude'] = pd.to_numeric(
C:\Users\Soham\AppData\Local\Temp\ipykernel_34712\1807732926.py:6: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_DF['longitude'] = pd.to_numeric(
C:\Users\Soham\AppData\Local\Temp\ipykernel_34712\1807732926.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value inste

,date,time_only,serial,sensor_name,status_current_fill_level,latitude,longitude
0,2021-05-15,09:01:49,FB1000001905AA12,Sensor-12,19,-37.802367,144.966167
1,2021-05-15,10:53:51,FB1000001905AA11,Sensor-11,23,-37.802133,144.966083
2,2021-05-16,02:06:58,FB1000001905AA07,Sensor-07,5,-37.803117,144.965233


In [20]:
# Keep the latest available reading for each sensor
new_DF = (
    new_DF
    .sort_values(['date', 'time_only'])
    .drop_duplicates(subset='serial', keep='last')
    .reset_index(drop=True)
)

<div id="visualisation-part-1" class="usecase-section-heading"><b>3. Visualisation Part 1: Mapping Sensor Locations</b></div>

<div class="usecase-body">This section plots the smart-bin sensor locations on an interactive map.</div>

<div id="Visualising-Geolocation-Data" class="usecase-sub-section-heading"><b>3.1 Visualising Geolocation Data with Folium</b></div>

<div class="usecase-body">
In our project, we leverage the Folium library to create interactive maps that can visualise the geolocation data we've prepared. The first step involves installing Folium, a powerful Python library used for visualising geospatial data. Here’s how we proceed:
</div>

<div class="usecase-list">

- **Installation**: We begin by installing Folium using `pip install folium`, ensuring that we have access to its comprehensive mapping capabilities.

- **Importing Folium**: With Folium installed, we import it into our notebook to start creating maps.

- **Setting Coordinates**: We define the coordinates for Argyle Square, which will serve as the focal point of our map. The latitude (`-37.8036`) and longitude (`144.9655`) values are set to centre the map precisely on Argyle Square.

- **Creating the Map**: We instantiate a Folium map object using `folium.Map()`, passing in the coordinates of Argyle Square along with a `zoom_start` value of 17 to provide a detailed view of the area. The `tiles` parameter is set to 'CartoDB positron' for a clean and modern map aesthetic.
</div>

<div class="usecase-body"> This step is crucial for visualising the spatial distribution of our data points, offering insights into patterns and trends that may not be apparent through numerical analysis alone. The interactive nature of Folium maps further enhances our ability to explore and interpret the data effectively. </div>



In [21]:
%pip install folium

Note: you may need to restart the kernel to use updated packages.


In [ ]:
# Create the Map Centred on Argyle Square 
# Coordinates for Argyle Square
argyle_square_lat = -37.8036
argyle_square_lon = 144.9655
 
# Create a map centred around Argyle Square with a suitable zoom level
map = folium.Map(location=[argyle_square_lat, argyle_square_lon], zoom_start=17, tiles='CartoDB positron')

<div id="Adding-Markers-to-the-Map" class="usecase-sub-section-heading"><b>3.2 Adding Markers to the Map for Each Location</b></div>

<div class="usecase-body">
With our Folium map centred on Argyle Square, the next step enhances our spatial analysis by marking the precise locations of interest within our dataset. Here’s how we systematically add markers to our map:
</div>
<div class="usecase-list">

- **Iterating Through Data**: We loop through each row in our `new_DF` DataFrame, which contains the cleaned and preprocessed data, including the essential 'latitude' and 'longitude' columns for our geolocation points.

- **Marker Creation**: For each row, we create a Folium marker using `folium.Marker()`. This marker is set to the geographical coordinates found in the 'latitude' and 'longitude' columns of the row. 

- **Popup Information**: To each marker, we add a popup that displays the latitude and longitude of the location. This popup is activated when the marker is clicked, providing quick access to precise location data.

- **Adding Markers to the Map**: Each marker is then added to our previously created Folium map object with `.add_to(map)`. This action places the marker on the map at the designated coordinates.
</div>

<div class="usecase-body"> By marking each location on the map, we visually represent our dataset's geospatial distribution. This method not only aids in identifying patterns or clusters in the data but also facilitates a more interactive and engaging analysis, allowing users to explore individual data points through their geographic context. </div>



In [23]:
# Add Markers for All Latitude and Longitude Values
for index, row in new_DF.iterrows():
    folium.Marker(
        [row['latitude'], row['longitude']],
        popup=f"Lat: {row['latitude']}, Lon: {row['longitude']}"
    ).add_to(map)

<div id="Visualising-Sensor-Locations-on-a-Map" class="usecase-sub-section-heading"><b>3.3 Visualising Sensor Locations on a Map</b></div>

<div class="usecase-body">
To further enhance our spatial analysis, we create a simplified map centred around Argyle Square, using it as a base to display the locations of various sensors. This visual representation is key to understanding the geographical distribution of sensors and assessing any spatial patterns. Here's the step-by-step process: </div>

<div class="usecase-list">

- **Map Creation**: We start by generating a new Folium map object, `simple_map`, centred on Argyle Square. The map is initialised with a zoom level of 17, providing a close-up view of the area for detailed inspection.

- **Adding Markers**: We iterate over our dataset, adding a marker for each sensor's location. Each marker is configured with the sensor's latitude and longitude, ensuring accurate placement on the map.

- **Popup Feature**: For added interactivity, each marker is equipped with a popup that displays the sensor's latitude and longitude. This feature allows users to click on any marker to reveal the exact coordinates of the sensor, enhancing the map's informational value.

- **Displaying the Map**: Finally, the `simple_map` object is displayed, showing all sensor locations as markers. When using a Jupyter notebook, the map will automatically render within the notebook interface, providing an immediate and interactive visual analysis tool.
</div>

<div class="usecase-body"> By incorporating these steps, we create a comprehensive spatial visualisation that allows for an intuitive exploration of sensor data, aiding in the identification of patterns, clusters, or gaps in sensor coverage. </div>


In [ ]:
# # Create a map centred around the average location
# Bin_location_map = folium.Map(location=[argyle_square_lat, argyle_square_lon], zoom_start=17)
 
# # Loop through the DataFrame and add each bench location as a marker on the map
# for index, row in new_DF.iterrows():
#     folium.Marker(
#         location=[row['latitude'], row['longitude']],
#         popup=f"Lat: {row['latitude']}, Lon: {row['longitude']}"
#     ).add_to(Bin_location_map)
 
# # To display the map in a Jupyter notebook
# # Bin_location_map
# display(Bin_location_map)

# Create a map centred around Argyle Square
Bin_location_map = folium.Map(location=[argyle_square_lat, argyle_square_lon], zoom_start=17, tiles='https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tile/{z}/{y}/{x}',
    attr='Tiles © Esri')

# Add smart-bin locations using the template green
for index, row in new_DF.iterrows():
    folium.CircleMarker(
        location=[row['latitude'], row['longitude']], radius=6, color='#059669', fill=True, 
        fill_color='#059669', fill_opacity=0.8, popup=f"Lat: {row['latitude']}, Lon: {row['longitude']}"
        ).add_to(Bin_location_map)

# Legend
legend_html = """
<div style="
position: fixed;
bottom: 40px;
left: 40px;
background-color: white;
padding: 8px;
z-index: 9999;">
<b>Legend</b><br>
<span style="color:#059669;">●</span> Smart Bin Location
</div>
"""

Bin_location_map.get_root().html.add_child(folium.Element(legend_html))

from IPython.display import display
display(Bin_location_map)

<div id="visualisation-part-2" class="usecase-section-heading"><b>4. Visualisation Part 2: Finding the Best Route</b></div>

<div class="usecase-body">This section uses bin fill levels and location data to determine an efficient collection route.</div>

<div id="Preparing-Fill-Levels" class="usecase-sub-section-heading"><b>4.1 Preparing Fill Levels for Route Prioritisation</b></div>

<div class="usecase-body">
The `status_current_fill_level` values are already represented as percentages from 0 to 100, so no additional normalisation is required.

These original fill-level percentages are used directly to rank the smart bins, with higher fill levels receiving greater collection priority during route planning.
</div>

<div id="Calculating-Distance-Matrix" class="usecase-sub-section-heading"><b>4.2 Calculating the Pairwise Distance Matrix</b></div>
<div class="usecase-body">
To compare the locations of the smart bins, the latitude and longitude coordinates are first converted from degrees to radians.

The Haversine method is then used to calculate the approximate great-circle distance between every pair of bin locations. This is more suitable than applying Euclidean distance directly to latitude and longitude because geographic coordinates represent positions on the Earth's curved surface.

The resulting `dist_matrix` stores the pairwise distances in kilometres and is used in the following route-planning steps.
</div>

In [26]:

# Convert latitude and longitude from degrees to radians
coords_radians = np.radians(new_DF[['latitude', 'longitude']].to_numpy())

# Calculate pairwise Haversine distances
earth_radius_km = 6371.0

dist_matrix = (haversine_distances(coords_radians) * earth_radius_km)

print(np.round(dist_matrix, 3))

[[0.    0.104 0.006 0.02  0.082 0.034 0.067 0.058 0.08  0.113 0.03  0.065
  0.093 0.054 0.045 0.091]
 [0.104 0.    0.098 0.091 0.143 0.107 0.078 0.162 0.125 0.117 0.132 0.084
  0.012 0.149 0.098 0.019]
 [0.006 0.098 0.    0.016 0.083 0.034 0.064 0.064 0.079 0.111 0.035 0.061
  0.088 0.058 0.044 0.085]
 [0.02  0.091 0.016 0.    0.072 0.05  0.071 0.074 0.066 0.121 0.041 0.045
  0.079 0.059 0.057 0.076]
 [0.082 0.143 0.083 0.072 0.    0.116 0.143 0.091 0.022 0.193 0.07  0.06
  0.131 0.053 0.127 0.125]
 [0.034 0.107 0.034 0.05  0.116 0.    0.044 0.071 0.114 0.082 0.055 0.093
  0.098 0.082 0.015 0.098]
 [0.067 0.078 0.064 0.071 0.143 0.044 0.    0.115 0.135 0.052 0.095 0.103
  0.073 0.121 0.03  0.077]
 [0.058 0.162 0.064 0.074 0.091 0.071 0.115 0.    0.103 0.148 0.033 0.109
  0.151 0.038 0.086 0.149]
 [0.08  0.125 0.079 0.066 0.022 0.114 0.135 0.103 0.    0.186 0.077 0.041
  0.113 0.066 0.122 0.106]
 [0.113 0.117 0.111 0.121 0.193 0.082 0.052 0.148 0.186 0.    0.137 0.155
  0.116 0.164 0.06

<div id="Route-Optimisation" class="usecase-sub-section-heading"><b>4.3 Route Optimisation Based on Fill Levels and Proximity</b></div>

<div class="usecase-body">
To enhance operational efficiency in servicing bins, we employ a strategy that prioritises bins with higher fill levels and minimises travel distance. This two-pronged approach ensures that the most critical bins are serviced first while also optimising the route taken to reduce time and resources spent. Here’s how we implement this strategy: </div>

<div id="Prioritising-by-Fill-Level" class="usecase-sub-sub-section-heading"><b> 4.3.1 Prioritising by Fill Level</b></div>

<div class="usecase-list">

- **Sorting Locations**: We begin by sorting the locations based on their fill levels in descending order. This is achieved by using `np.argsort` on the negative values of the `status_current_fill_level` column in `new_DF`. The result, `priority_indices`, represents the indices of locations sorted by their priority, with higher fill levels first.</div>


In [27]:
# Sort locations by descending fill level (prioritize higher fill levels)
new_DF['status_current_fill_level'] = pd.to_numeric(new_DF['status_current_fill_level'],errors='coerce')

new_DF = new_DF.dropna(subset=['status_current_fill_level'])

priority_indices = np.argsort(-new_DF['status_current_fill_level'].to_numpy())

<div id="Route-Initialisation" class="usecase-sub-sub-section-heading"><b>4.3.2 Route Initialisation</b></div>
<div class="usecase-list">

- **Initialising the Route**: The route list is initialised with the starting point, chosen as the first location in our prioritised list. This represents the location with the highest fill level.
</div>

In [28]:
# Initialize route list with the starting point (assuming first location as start)
route = [priority_indices[0]]

<div id="Determine-Next-Stop" class="usecase-sub-sub-section-heading"><b>4.3.3 Determine Next Stop</b></div>

<div class="usecase-body"> To determine the next stop, we follow a simple heuristic: from the current location, move to the nearest unvisited location. For each subsequent location: </div>

<div class="usecase-list">

- Calculate the distances from the last visited location to all others using the precomputed distance matrix, dist_matrix.

- Temporarily set the distance to already visited locations to infinity to exclude them from consideration.

- Identify the next stop as the location with the minimum distance to the last visited location. This location is then added to the route.
</div>

<div class="usecase-body"> This approach prioritises the bin with the highest fill level and then selects the nearest unvisited locations. This helps reduce unnecessary travel distance while ensuring the highest-priority bin is serviced first.</div>

In [29]:
for _ in range(1, len(priority_indices)):
    last_visited = route[-1]
    distances_to_last_visited = dist_matrix[last_visited].copy()
    # Filter out already visited locations
    distances_to_last_visited[route] = np.inf
    next_stop = np.argmin(distances_to_last_visited)
    route.append(next_stop)


In [30]:
# Calculate the total distance of the proposed route
total_route_distance = 0

for i in range(len(route) - 1):
    total_route_distance += dist_matrix[route[i], route[i + 1]]

print(f"Total proposed route distance: {total_route_distance:.3f} km")

Total proposed route distance: 0.591 km


<div class="usecase-body">
The total distance of the proposed collection sequence is calculated using the Haversine distances between consecutive stops. This provides a simple quantitative measure for evaluating the generated route rather than relying only on the map visualisation.</div>

In [31]:
# Calculate a baseline route using fill-level priority order
baseline_route = list(priority_indices)

baseline_distance = 0

for i in range(len(baseline_route) - 1):
    baseline_distance += dist_matrix[
        baseline_route[i],
        baseline_route[i + 1]
    ]

print(f"Baseline route distance: {baseline_distance:.3f} km")
print(f"Proposed route distance: {total_route_distance:.3f} km")

if baseline_distance > 0:
    improvement = (
        (baseline_distance - total_route_distance)
        / baseline_distance
    ) * 100

    print(f"Distance reduction: {improvement:.2f}%")

Baseline route distance: 1.229 km
Proposed route distance: 0.591 km
Distance reduction: 51.92%


<div class="usecase-body">
The proposed route reduces the total estimated travel distance from 1.229 km to 0.591 km, representing a distance reduction of approximately 51.92% compared with the fill-level-only baseline route.

For comparison, a baseline route follows the bins in fill-level priority order without considering proximity between consecutive stops. The proposed route is compared against this baseline using total Haversine distance. This provides a simple way to evaluate whether considering proximity reduces the overall collection distance.
</div>

<div id="Visualising-the-Optimised-Route" class="usecase-sub-section-heading"><b>4.4 Visualisation of the Optimised Route</b></div>

<div class="usecase-body">
To further refine our visualisation and provide a comprehensive view of the optimised route, we employ Folium to map out each stop with markers and visually connect these stops with lines. This approach not only highlights the individual locations but also delineates the sequence and path of the route, offering a holistic view of the journey. Here's the step-by-step process:
</div>
<div id="Setting-Up-the-Map" class="usecase-sub-sub-section-heading"><b>4.4.1 Setting Up the Map</b></div>
<div class="usecase-list">

- **Centring the Map**: We calculate the central point of our route by averaging the latitude and longitude of all stops. This ensures the map is optimally centred, making all route points easily viewable.</div>

In [32]:
# Visualisation
map_center = [new_DF['latitude'].mean(), new_DF['longitude'].mean()]
m = folium.Map(location=map_center, zoom_start=20, tiles='https://server.arcgisonline.com/ArcGIS/rest/services/World_Street_Map/MapServer/tile/{z}/{y}/{x}',
    attr='Tiles © Esri')

<div id="Marking-Stops" class="usecase-sub-sub-section-heading"><b>4.4.2 Marking Stops</b></div>
<div class="usecase-list">

- **Placing Marker**: Each location along the route is marked using a colour-coded marker based on its fill level. The markers include popups displaying the fill-level percentage, providing immediate insight into the priority of each stop.
</div>

In [33]:
# Add a marker for the starting point
start_idx = route[0]

folium.CircleMarker(
    location=[
        new_DF.iloc[start_idx]['latitude'],
        new_DF.iloc[start_idx]['longitude']
    ],
    radius=8,
    color='#9333EA',
    fill=True,
    fill_color='#9333EA',
    fill_opacity=1,
    popup='Start'
).add_to(m)


# Add a marker for the end point
end_idx = route[-1]

folium.CircleMarker(
    location=[
        new_DF.iloc[end_idx]['latitude'],
        new_DF.iloc[end_idx]['longitude']
    ],
    radius=8,
    color='#2563EB',
    fill=True,
    fill_color='#2563EB',
    fill_opacity=1,
    popup='End'
).add_to(m)


def get_marker_color(fill_level):
    """Return a template colour based on the smart-bin fill level."""
    if fill_level <= 30:
        return '#059669'   # Primary Green
    elif fill_level <= 70:
        return '#EAB308'   # Distinct Yellow
    else:
        return '#F97316'   # Distinct Orange


for idx in route:
    fill_level = new_DF.iloc[idx]['status_current_fill_level']
    marker_colour = get_marker_color(fill_level)

    folium.CircleMarker(
        location=[
            new_DF.iloc[idx]['latitude'],
            new_DF.iloc[idx]['longitude']
        ],
        radius=6,
        color=marker_colour,
        fill=True,
        fill_color=marker_colour,
        fill_opacity=0.9,
        popup=f'Fill Level: {fill_level}%'
    ).add_to(m)

In [34]:
print("Rows in new_DF:", len(new_DF))
print("Locations in route:", len(route))
print("Unique sensors:", new_DF['serial'].nunique())

Rows in new_DF: 16
Locations in route: 16
Unique sensors: 16


<div id="Illustrating-the-Route" class="usecase-sub-sub-section-heading"><b>4.4.3 Illustrating the Route</b></div>
<div class="usecase-list">

- **Connecting with Lines**: To visualise the travel path, green lines are drawn between consecutive stops along the optimised route. This is achieved by iterating through the route and adding a polyline between each pair of selected smart-bin locations, making the collection sequence easier to follow on the map.

</div>

In [35]:
# # Display route as lines
# for i in range(len(route) - 1):
#     start_pos = [new_DF.iloc[route[i]]['latitude'],new_DF.iloc[route[i]]['longitude']]
#     end_pos = [new_DF.iloc[route[i+1]]['latitude'], new_DF.iloc[route[i+1]]['longitude']]
#     folium.PolyLine([start_pos, end_pos], color='red').add_to(m)

# # To display the map in a Jupyter notebook
# m

########

# Display route as lines using the template green
for i in range(len(route) - 1):
    start_pos = [new_DF.iloc[route[i]]['latitude'], new_DF.iloc[route[i]]['longitude']]

    end_pos = [new_DF.iloc[route[i + 1]]['latitude'], new_DF.iloc[route[i + 1]]['longitude']]

    folium.PolyLine([start_pos, end_pos], color='#059669', weight=4).add_to(m)

# Add legend
legend_html = """
<div style="
position: fixed;
bottom: 40px;
left: 40px;
background-color: white;
padding: 8px;
z-index: 9999;">
<b>Legend</b><br>
<span style="color:#059669;">━</span> Optimised Route<br>
<span style="color:#059669;">●</span> Fill ≤ 30%<br>
<span style="color:#EAB308;">●</span> Fill 31–70%<br>
<span style="color:#F97316;">●</span> Fill > 70%<br>
<span style="color:#9333EA;">●</span> Start<br>
<span style="color:#2563EB;">●</span> End
</div>
"""

m.get_root().html.add_child(
    folium.Element(legend_html)
)

m

<div id="conclusion" class="usecase-unnumbered-heading"><b>Conclusion</b></div>


<div class="usecase-body">The use case addressed the waste collection problem defined in the Scenario by using smart-bin fill levels and geographic locations to develop a more efficient collection route. The proposed approach prioritised the bin with the highest fill level and then selected nearby unvisited bins to reduce unnecessary travel. Compared with the fill-priority baseline route, the proposed route reduced the estimated travel distance from approximately 1.229 km to 0.591 km.</div>

<div class="usecase-body"><b>What we achieved in this analysis</b></div>

<div class="usecase-body">
This analysis used Melbourne smart-bin sensor data to examine bin fill levels and geographic locations and develop a more efficient waste collection route. Haversine distance was used to calculate realistic geographic distances between bin locations, and the route began with the bin having the highest fill level before selecting the nearest unvisited locations. The proposed route had an estimated total distance of approximately 0.591 km compared with 1.229 km for the fill-priority baseline route, representing a reduction of approximately 51.92%. This result is significant because it shows that considering geographic proximity, rather than visiting bins only according to fill-level order, can substantially reduce unnecessary travel while still servicing the highest-priority bin first. For waste collection planning, a shorter route could support more efficient use of collection vehicles and reduce the amount of travel required between stops.

</div>

<div class="usecase-body"><b>What we learned from this analysis</b></div>

<div class="usecase-body">
The analysis demonstrates that combining smart-bin fill-level information with geographic distance can support more efficient waste collection planning. The visualisations also make it easier to understand where the bins are located, their current fill levels and the sequence in which they could be serviced. Comparing the proposed route with a baseline route showed that considering proximity after prioritising the fullest bin can substantially reduce unnecessary travel distance.
</div>

<div class="usecase-body"><b>Observations for further opportunities</b></div>

<div class="usecase-body">
The current analysis is based on one smart-bin dataset and focuses mainly on bin fill levels and geographic distance. Future development could incorporate additional factors such as real-time traffic conditions, vehicle capacity, collection time windows and changing bin fill levels throughout the day. These additional data sources could support more dynamic route planning and further improve operational efficiency while reducing fuel consumption and unnecessary travel across Melbourne.
</div>